web/packages/teleterm/src/ui/DocumentGatewayKube/DocumentGatewayKube.tsx

```diff
@@ -0,0 +1,160 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import React, { useState, useEffect } from 'react';
+import styled from 'styled-components';
+import { Flex, Text, ButtonPrimary } from 'design';
+import { useAsync } from 'shared/hooks/useAsync';
+
+import Document from 'teleterm/ui/Document';
+import * as types from 'teleterm/ui/services/workspacesService';
+import { useAppContext } from 'teleterm/ui/appContextProvider';
+import { useWorkspaceContext } from 'teleterm/ui/Documents';
+import { DocumentTerminal } from 'teleterm/ui/DocumentTerminal';
+import { connectToKube } from 'teleterm/ui/services/workspacesService';
+import { retryWithRelogin } from 'teleterm/ui/utils';
+import { routing } from 'teleterm/ui/uri';
+
+export const DocumentGatewayKube = (props: {
+  visible: boolean;
+  doc: types.DocumentGatewayKube;
+}) => {
+  const { clustersService } = useAppContext();
+  clustersService.useState();
+
+  const { doc, visible } = props;
+  const [hasRenderedTerminal, setHasRenderedTerminal] = useState(false);
+
+  // TODO support user, groups, namespace
+  const gateway = clustersService.findGatewayByConnectionParams(
+    doc.targetUri,
+    ''
+  );
+
+  if (gateway || hasRenderedTerminal) {
+    if (!hasRenderedTerminal) {
+      setHasRenderedTerminal(true);
+    }
+
+    return <DocumentTerminal doc={doc} visible={visible} />;
+  }
+
+  return <WaitingForGateway doc={doc} visible={visible} />;
+};
+
+const TIMEOUT_SECONDS = 10;
+
+const WaitingForGateway = (props: {
+  doc: types.DocumentGatewayKube;
+  visible: boolean;
+}) => {
+  const { doc, visible } = props;
+  const ctx = useAppContext();
+  const { documentsService } = useWorkspaceContext();
+  // If we depended on doc.status for hasTimedOut instead of using a separate state, then on reopen
+  // the doc would have status set to 'connected' on 'error' and it'd be updated from useEffect,
+  // meaning that there would be a brief flash of old state.
+  const [hasTimedOut, setHasTimedOut] = useState(false);
+  const [connectAttempt, createGateway] = useAsync(async () => {
+    const gw = await retryWithRelogin(ctx, doc.targetUri, () =>
+      // TODO support user, groups, namespace
+      ctx.clustersService.createGateway({
+        targetUri: doc.targetUri,
+        user: '',
+      })
+    );
+
+    documentsService.update(doc.uri, {
+      gatewayUri: gw.uri,
+      port: gw.localPort,
+    });
+  });
+  const { params } = routing.parseKubeUri(doc.targetUri);
+
+  useEffect(() => {
+    // Update the doc state to make the progress bar show up in the tab bar.
+    // Once DocumentTerminal is mounted, it is going to update the status to 'connected' or 'error'.
+    documentsService.update(doc.uri, { status: 'connecting' });
+
+    if (connectAttempt.status === '') {
+      createGateway();
+    }
```

Comment:
Unlike `DocumentGatewayCliClient`, `DocumentGatewayKube` is responsible for creating the gateway, so if an error happens, we have to remember about updating the status of the doc.

```suggestion
      (async () => {
        const [, error] = await createGateway();

        if (!(error instanceof CanceledError)) {
          documentsService.update(doc.uri, { status: 'error' });
        }
      })();
```

`CanceledError` is from `shared/hooks/useAsync`.